pkg/upgrade/upgradecluster/tenant_cluster.go

```diff
@@ -120,38 +129,162 @@ import (
+	// function as transactional protection is required by the tenant upgrade
+	// interlock.
+	instances, err := t.InstanceReader.GetAllInstancesNoCache(ctx)
+	if err != nil {
+		return err
+	}
+
+	// If we're bumping the cluster version, cache the list of instances we
+	// contacted at bump time. This list of instances is then consulted at
+	// fence write time to ensure that we contacted the same set of instances.
+	if strings.Contains(op, BumpClusterVersionOpName) {
+		t.instancesAtBump = instances
+	}
+
+	// Limiting of outgoing RPCs at the tenant level mirrors what we do for
+	// nodes at the storage cluster level.
+	qp := quotapool.NewIntPool("every-sql-pod", 25)
+	log.Infof(ctx, "executing %s on nodes %v", redact.Safe(op), instances)
+	grp := ctxgroup.WithContext(ctx)
+
+	for _, instance := range instances {
+		id := instance.InstanceID // copy out of the loop variable
+		alloc, err := qp.Acquire(ctx, 1)
+		if err != nil {
+			return err
+		}
+
+		grp.GoCtx(func(ctx context.Context) error {
+			defer alloc.Release()
+
+			conn, err := t.Dialer.Dial(ctx, roachpb.NodeID(id), rpc.DefaultClass)
+			if err != nil {
+				if strings.Contains(err.Error(), "failed to connect") {
+					if strings.Contains(err.Error(), "version compatibility check failed") {
+						return errors.WithHint(errors.Newf("upgrade failed due to active SQL servers with incompatible binary version(s)"),
+							"upgrade the binary versions of all SQL servers before re-attempting the tenant upgrade")
+					} else {
+						return errors.WithHint(errors.Newf("upgrade failed due to transient SQL servers"),
+							"retry upgrade when the SQL servers for the given tenant are in a stable state (i.e. not starting/stopping)")
+					}
+				}
+				return err
+			}
+			client := serverpb.NewMigrationClient(conn)
+			return fn(ctx, client)
+		})
+	}
+	return grp.Wait()
 }
 
 // UntilClusterStable is part of the upgrade.Cluster interface.
 //
-// Tenant clusters in the current version assume their cluster is stable
-// because they presently assume there is at most one running SQL pod. When
-// that changes, this logic will need to change.
-func (t TenantCluster) UntilClusterStable(ctx context.Context, fn func() error) error {
+// We don't have the same notion of cluster stability with tenant pods as we
+// do with cluster nodes. As a result, this function behaves slightly
+// differently with secondary tenants than it does with the system tenant.
+// Instead of relying on liveness and waiting until all nodes are active before
+// we claim that the cluster has become "stable", we test to see how many
+// pods of the tenant are active, and loop until we find two successive
+// iterations where the number of active tenant pods is the same.
+func (t *TenantCluster) UntilClusterStable(ctx context.Context, fn func() error) error {
+	instances, err := t.InstanceReader.GetAllInstancesNoCache(ctx)
+	if err != nil {
+		return err
+	}
+
+	for {
```

Comment:
This feels like something that should use a retrier and back off to at least a second